qmi/core/transport.py

```diff
@@ -612,155 +635,314 @@ def _safe_socket(self) -> socket.socket:
+        raise NotImplementedError("QMI_UdpTcpTransportBase.write not implemented")
+
+    def read(self, nbytes: int, timeout: Optional[float] = None) -> bytes:
+        raise NotImplementedError("QMI_UdpTcpTransportBase.read not implemented")
+
+    def read_until(self, message_terminator: bytes, timeout: Optional[float] = None) -> bytes:
+        # Check if message terminator already received. To be further handled in subclass implementation.
+        p = self._read_buffer.find(message_terminator)
+        if p >= 0:
+            # Found "message_terminator" - return data.
+            nbytes = p + len(message_terminator)
+            ret = bytes(self._read_buffer[:nbytes])
+            self._read_buffer = self._read_buffer[nbytes:]
+            return ret
+
+        return b""
+
+    def read_until_timeout(self, nbytes: int, timeout: float) -> bytes:
+        try:
+            ret = self.read(nbytes, timeout)
+        except QMI_TimeoutException:
+            ret = bytes(self._read_buffer)
+            self._read_buffer = bytearray()
+        except QMI_EndOfInputException:
+            if not self._read_buffer:
+                raise
+            ret = bytes(self._read_buffer)
+            self._read_buffer = bytearray()
+        return ret
 
+    def discard_read(self) -> None:
+        self._check_is_open()
         self._read_buffer = bytearray()
+        self._safe_socket.settimeout(0)
+        while True:
+            try:
+                b, addr = self._safe_socket.recvfrom(4096)
+            except (BlockingIOError, socket.timeout):
+                # no more bytes available
+                break
+            except OSError:
+                # UDP protocol was used and > 4096 bytes in socket buffer. This discards also the rest of the packet.
+                break
+
+            if not b:
+                # end of stream
+                break
+
+
+class QMI_UdpTransport(QMI_UdpTcpTransportBase):
+    """
+    An instance of QMI_UdpTransport represents a server-side UDP connection with
+    a listening port to an instrument. Client-side UDP connections are not supported.
+
+    The maximum UDP packet size is 64KB (including headers). For `write` and `read` functions,
+    this means limitations. The write function we can check the data size to send and split into
+    multiple messages. For `read` functions this is not possible and the client must limit the
+    data size as necessary. If for `read` functions the receivable data size is larger than
+    the size defined in `recvfrom(<size>)`, an exception is raised.
+    """
+
+    def __init__(self, host: str, port: int) -> None:
+        """Initialize the UDP transport by validating the UDP host and port.
+
+        Parameters:
+            host: The server IP address.
+            port: The port for the address.
+        """
+        self._validate_udp_port(port)
+        super().__init__(host, port)
+
+    @staticmethod
+    def _validate_udp_port(port):
```

Comment:
minor: why not just add this `self._validate_port(...)` method?
```python
    @staticmethod
    def _validate_port(port):
        super()._validate_port(port)
        if port == QMI_Context.DEFAULT_UDP_RESPONDER_PORT:
            raise QMI_TransportDescriptorException(f"UDP port number {port} not allowed")
```

Reply:
Why not? Thanks for the tip, will change this!